File: flut/.flutter/lib/flutter/widgets/navigator.dart
import 'package:flutter/material.dart';
import 'package:flut/flut/object.dart';
import 'package:flut/flut/runtime.dart';
import 'package:flut/flut/error.dart';

class FlutRouteSettings extends FlutValueObject {
  final RouteSettings settings;
  const FlutRouteSettings(this.settings) : super('RouteSettings');

  @override
  Map<String, dynamic> flutEncode() {
    final result = flutBaseProps();
    if (settings.name != null) result['name'] = settings.name;
    if (settings.arguments != null) result['arguments'] = settings.arguments;
    return result;
  }

  static RouteSettings? flutDecode(
    FlutRuntime runtime,
    Map<String, dynamic> data,
  ) {
    return RouteSettings(
      name: runtime.unpackOptionalField<String>(data, 'name'),
      arguments: runtime.unpackOptionalField<Object>(data, 'arguments'),
    );
  }
}

class FlutNavigator {
  FlutNavigator._();

  static void registerStatics(FlutRuntime runtime) {
    runtime.registerStatic('Navigator.of', _of);
    runtime.registerStatic('Navigator.maybeOf', _maybeOf);
    runtime.registerStatic('Navigator.push', _push);
    runtime.registerStatic('Navigator.pushReplacement', _pushReplacement);
    runtime.registerStatic('Navigator.pushAndRemoveUntil', _pushAndRemoveUntil);
    runtime.registerStatic('Navigator.pushNamed', _pushNamed);
    runtime.registerStatic(
      'Navigator.pushReplacementNamed',
      _pushReplacementNamed,
    );
    runtime.registerStatic(
      'Navigator.pushNamedAndRemoveUntil',
      _pushNamedAndRemoveUntil,
    );
    runtime.registerStatic('Navigator.popAndPushNamed', _popAndPushNamed);
    runtime.registerStatic('Navigator.pop', _pop);
    runtime.registerStatic('Navigator.maybePop', _maybePop);
    runtime.registerStatic('Navigator.popUntil', _popUntil);
    runtime.registerStatic('Navigator.canPop', _canPop);
    runtime.registerStatic('Navigator.replace', _replace);
    runtime.registerStatic('Navigator.replaceRouteBelow', _replaceRouteBelow);
    runtime.registerStatic('Navigator.removeRoute', _removeRoute);
    runtime.registerStatic('Navigator.removeRouteBelow', _removeRouteBelow);
  }

  static dynamic _of(
    FlutRuntime runtime,
    BuildContext context, {
    bool rootNavigator = false,
  }) {
    return Navigator.of(context, rootNavigator: rootNavigator);
  }

  static dynamic _maybeOf(
    FlutRuntime runtime,
    BuildContext context, {
    bool rootNavigator = false,
  }) {
    return Navigator.maybeOf(context, rootNavigator: rootNavigator);
  }

  static dynamic _push(FlutRuntime runtime, BuildContext context, Route route) {
    return Navigator.push(context, route);
  }

  static dynamic _pushReplacement(
    FlutRuntime runtime,
    BuildContext context,
    Route newRoute, {
    dynamic result,
  }) {
    return Navigator.pushReplacement(context, newRoute, result: result);
  }

  static dynamic _pushAndRemoveUntil(
    FlutRuntime runtime,
    BuildContext context,
    Route newRoute,
    FlutCallableRef predicate,
  ) {
    return Navigator.pushAndRemoveUntil(
      context,
      newRoute,
      runtime.adaptCallableByType(predicate) as RoutePredicate,
    );
  }

  static dynamic _pushNamed(
    FlutRuntime runtime,
    BuildContext context,
    String routeName, {
    Object? arguments,
  }) {
    return Navigator.pushNamed(context, routeName, arguments: arguments);
  }

  static dynamic _pushReplacementNamed(
    FlutRuntime runtime,
    BuildContext context,
    String routeName, {
    dynamic result,
    Object? arguments,
  }) {
    return Navigator.pushReplacementNamed(
      context,
      routeName,
      result: result,
      arguments: arguments,
    );
  }

  static dynamic _pushNamedAndRemoveUntil(
    FlutRuntime runtime,
    BuildContext context,
    String newRouteName,
    FlutCallableRef predicate, {
    Object? arguments,
  }) {
    return Navigator.pushNamedAndRemoveUntil(
      context,
      newRouteName,
      runtime.adaptCallableByType(predicate) as RoutePredicate,
      arguments: arguments,
    );
  }

  static dynamic _popAndPushNamed(
    FlutRuntime runtime,
    BuildContext context,
    String routeName, {
    dynamic result,
    Object? arguments,
  }) {
    return Navigator.popAndPushNamed(
      context,
      routeName,
      result: result,
      arguments: arguments,
    );
  }

  static dynamic _pop(
    FlutRuntime runtime,
    BuildContext context, [
    dynamic result,
  ]) {
    Navigator.pop(context, result);
    return null;
  }

  static dynamic _maybePop(
    FlutRuntime runtime,
    BuildContext context, [
    dynamic result,
  ]) {
    return Navigator.maybePop(context, result);
  }

  static dynamic _popUntil(
    FlutRuntime runtime,
    BuildContext context,
    FlutCallableRef predicate,
  ) {
    Navigator.popUntil(
      context,
      runtime.adaptCallableByType(predicate) as RoutePredicate,
    );
    return null;
  }

  static dynamic _canPop(FlutRuntime runtime, BuildContext context) {
    return Navigator.canPop(context);
  }

  static dynamic _replace(
    FlutRuntime runtime,
    BuildContext context, {
    required Route oldRoute,
    required Route newRoute,
  }) {
    Navigator.replace(context, oldRoute: oldRoute, newRoute: newRoute);
    return null;
  }

  static dynamic _replaceRouteBelow(
    FlutRuntime runtime,
    BuildContext context, {
    required Route anchorRoute,
    required Route newRoute,
  }) {
    Navigator.replaceRouteBelow(
      context,
      anchorRoute: anchorRoute,
      newRoute: newRoute,
    );
    return null;
  }

  static dynamic _removeRoute(
    FlutRuntime runtime,
    BuildContext context,
    Route route,
  ) {
    Navigator.removeRoute(context, route);
    return null;
  }

  static dynamic _removeRouteBelow(
    FlutRuntime runtime,
    BuildContext context,
    Route anchorRoute,
  ) {
    Navigator.removeRouteBelow(context, anchorRoute);
    return null;
  }
}

class FlutNavigatorState with FlutRealtimeObject<NavigatorState> {
  FlutNavigatorState.createFromObject({
    required FlutRuntime runtime,
    required int oid,
    required NavigatorState target,
  }) {
    initRealtimeFromObject(
      runtime: runtime,
      oid: oid,
      type: 'NavigatorState',
      target: target,
    );
  }

  @override
  dynamic getRawProperty(String property) {
    throw FlutUnknownPropertyException('NavigatorState', property);
  }

  @override
  bool setProperty(String property, dynamic value) {
    throw FlutUnknownPropertyException('NavigatorState', property);
  }

  @override
  dynamic callMethod(
    String method,
    List<dynamic> args,
    Map<String, dynamic> kwargs,
  ) {
    switch (method) {
      case 'push':
        return flutTarget.push(args[0] as Route);
      case 'pushReplacement':
        return flutTarget.pushReplacement(
          args[0] as Route,
          result: kwargs['result'],
        );
      case 'pushAndRemoveUntil':
        return flutTarget.pushAndRemoveUntil(
          args[0] as Route,
          runtime.adaptCallableByType(args[1] as FlutCallableRef)
              as RoutePredicate,
        );
      case 'pushNamed':
        return flutTarget.pushNamed(
          args[0] as String,
          arguments: kwargs['arguments'],
        );
      case 'pushReplacementNamed':
        return flutTarget.pushReplacementNamed(
          args[0] as String,
          result: kwargs['result'],
          arguments: kwargs['arguments'],
        );
      case 'pushNamedAndRemoveUntil':
        return flutTarget.pushNamedAndRemoveUntil(
          args[0] as String,
          runtime.adaptCallableByType(args[1] as FlutCallableRef)
              as RoutePredicate,
          arguments: kwargs['arguments'],
        );
      case 'popAndPushNamed':
        return flutTarget.popAndPushNamed(
          args[0] as String,
          result: kwargs['result'],
          arguments: kwargs['arguments'],
        );
      case 'pop':
        if (args.isEmpty) {
          flutTarget.pop();
        } else {
          flutTarget.pop(args[0]);
        }
        return null;
      case 'maybePop':
        if (args.isEmpty) {
          return flutTarget.maybePop();
        }
        return flutTarget.maybePop(args[0]);
      case 'popUntil':
        flutTarget.popUntil(
          runtime.adaptCallableByType(args[0] as FlutCallableRef)
              as RoutePredicate,
        );
        return null;
      case 'canPop':
        return flutTarget.canPop();
      case 'replace':
        flutTarget.replace(
          oldRoute: kwargs['oldRoute'] as Route,
          newRoute: kwargs['newRoute'] as Route,
        );
        return null;
      case 'replaceRouteBelow':
        flutTarget.replaceRouteBelow(
          anchorRoute: kwargs['anchorRoute'] as Route,
          newRoute: kwargs['newRoute'] as Route,
        );
        return null;
      case 'removeRoute':
        flutTarget.removeRoute(args[0] as Route);
        return null;
      case 'removeRouteBelow':
        flutTarget.removeRouteBelow(args[0] as Route);
        return null;
      case 'finalizeRoute':
        flutTarget.finalizeRoute(args[0] as Route);
        return null;
    }
    throw FlutUnknownMethodException(method);
  }
}

File: flut/flutter/widgets/navigator.py
from typing import Any, Callable, Optional, override

from flut._flut.engine import (
    FlutRealtimeObject,
    FlutValueObject,
    Future,
    call_dart_static,
    pack_static_callable,
    _flut_pack_value,
)
from flut._flut.runtime import _flut_unpack_optional_field


class RouteSettings(FlutValueObject):
    _flut_type = "RouteSettings"

    def __init__(self, *, name=None, arguments=None):
        super().__init__()
        self.name = name
        self.arguments = arguments

    @override
    def _flut_pack(self) -> dict:
        result = self._flut_base_props()
        if self.name is not None:
            result["name"] = _flut_pack_value(self.name)
        if self.arguments is not None:
            result["arguments"] = _flut_pack_value(self.arguments)
        return result

    @staticmethod
    def _flut_unpack(data: dict) -> "RouteSettings":
        return RouteSettings(
            name=_flut_unpack_optional_field(data, "name"),
            arguments=_flut_unpack_optional_field(data, "arguments"),
        )


class NavigatorState(FlutRealtimeObject):
    _flut_type = "NavigatorState"

    def push(self, route) -> Future:
        return self._flut_call("push", route)

    def pushReplacement(self, newRoute, *, result=None) -> Future:
        if result is not None:
            return self._flut_call("pushReplacement", newRoute, result=result)
        return self._flut_call("pushReplacement", newRoute)

    def pushAndRemoveUntil(self, newRoute, predicate: Callable[[Any], bool]) -> Future:
        return self._flut_call(
            "pushAndRemoveUntil",
            newRoute,
            self._register_action(predicate, "RoutePredicate"),
        )

    def pushNamed(self, routeName: str, *, arguments: Optional[Any] = None) -> Future:
        if arguments is not None:
            return self._flut_call("pushNamed", routeName, arguments=arguments)
        return self._flut_call("pushNamed", routeName)

    def pushReplacementNamed(
        self,
        routeName: str,
        *,
        result: Optional[Any] = None,
        arguments: Optional[Any] = None,
    ) -> Future:
        kwargs: dict[str, Any] = {}
        if result is not None:
            kwargs["result"] = result
        if arguments is not None:
            kwargs["arguments"] = arguments
        return self._flut_call("pushReplacementNamed", routeName, **kwargs)

    def pushNamedAndRemoveUntil(
        self,
        newRouteName: str,
        predicate: Callable[[Any], bool],
        *,
        arguments: Optional[Any] = None,
    ) -> Future:
        kwargs: dict[str, Any] = {}
        if arguments is not None:
            kwargs["arguments"] = arguments
        return self._flut_call(
            "pushNamedAndRemoveUntil",
            newRouteName,
            self._register_action(predicate, "RoutePredicate"),
            **kwargs,
        )

    def popAndPushNamed(
        self,
        routeName: str,
        *,
        result: Optional[Any] = None,
        arguments: Optional[Any] = None,
    ) -> Future:
        kwargs: dict[str, Any] = {}
        if result is not None:
            kwargs["result"] = result
        if arguments is not None:
            kwargs["arguments"] = arguments
        return self._flut_call("popAndPushNamed", routeName, **kwargs)

    def pop(self, result=None) -> None:
        if result is not None:
            self._flut_call("pop", result)
        else:
            self._flut_call("pop")

    def maybePop(self, result=None) -> Future:
        if result is not None:
            return self._flut_call("maybePop", result)
        return self._flut_call("maybePop")

    def popUntil(self, predicate: Callable[[Any], bool]) -> None:
        self._flut_call(
            "popUntil",
            self._register_action(predicate, "RoutePredicate"),
        )

    def canPop(self) -> bool:
        return bool(self._flut_call("canPop"))

    def replace(self, *, oldRoute, newRoute) -> None:
        self._flut_call("replace", oldRoute=oldRoute, newRoute=newRoute)

    def replaceRouteBelow(self, *, anchorRoute, newRoute) -> None:
        self._flut_call("replaceRouteBelow", anchorRoute=anchorRoute, newRoute=newRoute)

    def removeRoute(self, route) -> None:
        self._flut_call("removeRoute", route)

    def removeRouteBelow(self, anchorRoute) -> None:
        self._flut_call("removeRouteBelow", anchorRoute)

    def finalizeRoute(self, route) -> None:
        self._flut_call("finalizeRoute", route)


class Navigator:
    @staticmethod
    def of(context, *, rootNavigator: bool = False) -> NavigatorState:
        return call_dart_static(
            "Navigator", "of", context._flut_pack(), rootNavigator=rootNavigator
        )

    @staticmethod
    def maybeOf(context, *, rootNavigator: bool = False) -> Optional[NavigatorState]:
        return call_dart_static(
            "Navigator",
            "maybeOf",
            context._flut_pack(),
            rootNavigator=rootNavigator,
        )

    @staticmethod
    def push(context, route) -> Future:
        return call_dart_static(
            "Navigator", "push", context._flut_pack(), _flut_pack_value(route)
        )

    @staticmethod
    def pushReplacement(context, newRoute, *, result=None) -> Future:
        kwargs: dict[str, Any] = {}
        if result is not None:
            kwargs["result"] = _flut_pack_value(result)
        return call_dart_static(
            "Navigator",
            "pushReplacement",
            context._flut_pack(),
            _flut_pack_value(newRoute),
            **kwargs,
        )

    @staticmethod
    def pushAndRemoveUntil(
        context, newRoute, predicate: Callable[[Any], bool]
    ) -> Future:
        return call_dart_static(
            "Navigator",
            "pushAndRemoveUntil",
            context._flut_pack(),
            _flut_pack_value(newRoute),
            pack_static_callable(predicate, "RoutePredicate"),
        )

    @staticmethod
    def pushNamed(
        context, routeName: str, *, arguments: Optional[Any] = None
    ) -> Future:
        kwargs: dict[str, Any] = {}
        if arguments is not None:
            kwargs["arguments"] = _flut_pack_value(arguments)
        return call_dart_static(
            "Navigator", "pushNamed", context._flut_pack(), routeName, **kwargs
        )

    @staticmethod
    def pushReplacementNamed(
        context,
        routeName: str,
        *,
        result: Optional[Any] = None,
        arguments: Optional[Any] = None,
    ) -> Future:
        kwargs: dict[str, Any] = {}
        if result is not None:
            kwargs["result"] = _flut_pack_value(result)
        if arguments is not None:
            kwargs["arguments"] = _flut_pack_value(arguments)
        return call_dart_static(
            "Navigator",
            "pushReplacementNamed",
            context._flut_pack(),
            routeName,
            **kwargs,
        )

    @staticmethod
    def pushNamedAndRemoveUntil(
        context,
        newRouteName: str,
        predicate: Callable[[Any], bool],
        *,
        arguments: Optional[Any] = None,
    ) -> Future:
        kwargs: dict[str, Any] = {}
        if arguments is not None:
            kwargs["arguments"] = _flut_pack_value(arguments)
        return call_dart_static(
            "Navigator",
            "pushNamedAndRemoveUntil",
            context._flut_pack(),
            newRouteName,
            pack_static_callable(predicate, "RoutePredicate"),
            **kwargs,
        )

    @staticmethod
    def popAndPushNamed(
        context,
        routeName: str,
        *,
        result: Optional[Any] = None,
        arguments: Optional[Any] = None,
    ) -> Future:
        kwargs: dict[str, Any] = {}
        if result is not None:
            kwargs["result"] = _flut_pack_value(result)
        if arguments is not None:
            kwargs["arguments"] = _flut_pack_value(arguments)
        return call_dart_static(
            "Navigator",
            "popAndPushNamed",
            context._flut_pack(),
            routeName,
            **kwargs,
        )

    @staticmethod
    def pop(context, result=None) -> None:
        if result is not None:
            call_dart_static(
                "Navigator", "pop", context._flut_pack(), _flut_pack_value(result)
            )
        else:
            call_dart_static("Navigator", "pop", context._flut_pack())

    @staticmethod
    def maybePop(context, result=None) -> Future:
        if result is not None:
            return call_dart_static(
                "Navigator",
                "maybePop",
                context._flut_pack(),
                _flut_pack_value(result),
            )
        return call_dart_static("Navigator", "maybePop", context._flut_pack())

    @staticmethod
    def popUntil(context, predicate: Callable[[Any], bool]) -> None:
        call_dart_static(
            "Navigator",
            "popUntil",
            context._flut_pack(),
            pack_static_callable(predicate, "RoutePredicate"),
        )

    @staticmethod
    def canPop(context) -> bool:
        return bool(call_dart_static("Navigator", "canPop", context._flut_pack()))

    @staticmethod
    def replace(context, *, oldRoute, newRoute) -> None:
        call_dart_static(
            "Navigator",
            "replace",
            context._flut_pack(),
            oldRoute=_flut_pack_value(oldRoute),
            newRoute=_flut_pack_value(newRoute),
        )

    @staticmethod
    def replaceRouteBelow(context, *, anchorRoute, newRoute) -> None:
        call_dart_static(
            "Navigator",
            "replaceRouteBelow",
            context._flut_pack(),
            anchorRoute=_flut_pack_value(anchorRoute),
            newRoute=_flut_pack_value(newRoute),
        )

    @staticmethod
    def removeRoute(context, route) -> None:
        call_dart_static(
            "Navigator",
            "removeRoute",
            context._flut_pack(),
            _flut_pack_value(route),
        )

    @staticmethod
    def removeRouteBelow(context, anchorRoute) -> None:
        call_dart_static(
            "Navigator",
            "removeRouteBelow",
            context._flut_pack(),
            _flut_pack_value(anchorRoute),
        )
